### ISSUE TITLE
flicker on workspace change with open window

### ISSUE THREAD

**Issue by adabru** [gergo-salyi/multibg-sway#7]
I have configured a bright wallpaper for workspace 2 and a dark wallpaper for workspace 6. When I open a window in workspace 2 in fullscreen (or any large window size) and then switch to workspace 6, the bright wallpaper of workspace 2 flashes up for a very short moment before the dark background of workspace 6 is shown.

I guess sway "minimizes" the window from workspace 2 and shows the background before multibg-sway has any chance to change it?

**Comment by gergo-salyi** [comment]
In short no, I don't think it is possible to eliminate that one frame flash happening sometimes.

Sway: `switches the workspace  ->  submits sway IPC socket message to us  AND  schedules rendering new frame`
multibg-sway: `receive sway IPC socket message  ->  wayland-commit background layer change`
Sway: `receives background layer wayland-commit  ->  schedules rendering new frame`

Sway tells us that the workspace changed but cannot block on us / wait for us with the rendering. So there is no guarantee that the workspace change and the wallpaper change will be rendered together atomically on one next frame. 

Sometimes our wayland-commit about the changes we made to background layer will not be processed before sway submits the rendering of a new frame with the changed workspace. 

I wrote `multibg-sway` because before that the method was to kill and restart `swaybg` on every workspace change with a new wallpaper. That had bad latency - this solution minimizes it but cannot eliminate it fully. 

It should be not more than 1 frame tough - is it?

**Comment by adabru** [comment]
It seems like 1 frame to me. Could also be 2 or 3, I can't judge it well.

Won't fix seems good enough for me.

You pointing out the order of events brought a thought to me. Currently I'm switching workspaces via keybindings. If multibg-sway could instruct sway to change the background after it's finished changing the background, I could redirect the keybinding to multibg-sway. Or if wayland has some form of notifying about a background change, I could use that to wait for the background change. I checked `swaymsg -t subscribe` but it doesn't have notifications about background changes.

**Comment by gergo-salyi** [comment]
That would just reverse the order of happenings and keep the problem I think. Instead of having a transient `new workspace + old wallpaper` we would have a transient `old workspace + new wallpaper` flicker, wouldn't we?

**Comment by zDEFz** [comment]
> I wrote `multibg-sway` because before that the method was to kill and restart `swaybg` on every workspace change with a new wallpaper. That had bad latency - this solution minimizes it but cannot eliminate it fully.
> 
> It should be not more than 1 frame tough - is it?

We can circumvent the issue by adding a floating window for each workspace.
All you need is to make the `app_id` match with the current workspace.
Then you can make a rule to bind it to the workspace.
That would mean each workspace has an instance.

**Comment by danieldg** [comment]
You can get lower latency if you initiate the workspace change from your application instead of reacting to it.  This would require something like:

- Create an IPC mechanism to your client and a command to send it (this could be `echo cmd | socat stdio UNIX-SENDTO:/path/to/socket` or just executing your normal binary with specific arguments)
- Changing all bindings that switch workspaces to instead `exec` the IPC client requesting the change
- On receipt of a request, prepare the new background frame, and then send the commit() and the sway IPC message right after each other

The end result would be that sway itself gets the new background and the workspace change in the same frame and so it should apply both in the next frame.  There's still a race, but it's much smaller.

The only way to eliminate the race completely is to add support for this to sway itself, perhaps by adding a custom wayland protocol (or just ipc) that can bind layer-shell surfaces to workspaces - and then creating one surface per workspace, letting sway do the switching.

**Comment by gergo-salyi** [comment]
Good idea, I'm interested. 

> You can get lower latency if you initiate the workspace change from your application instead of reacting to it.

Just to clarify, @danieldg is this _speculation_ - or do you _know this for sure_ (eg. from sway's source), that in sway for `sway ipc -> SUBSCRIBE` the `sway ipc -> EVENTS -> workspace` events (that we are reacting for now) are getting sent _so prohibitively late_ (eg. after already rendering or such) during sway's internal processing of a workspace change?

If this is for sure then in the next 1-2 weeks I'll spend time to code a setup here where:

In `~/.config/sway/config`
```
bindsym $mod+1 nop
bindsym $mod+2 nop
# ...
bindsym $mod+0 nop
```
In our wallpaper app:
- Listen for these as sway ipc -> SUBSCRIBE -> EVENTS -> binding
- Respond with:
  - sway ipc -> RUN_COMMAND -> "workspace 1"
  - wayland surface commit()

**Comment by danieldg** [comment]
> Just to clarify, @danieldg is this speculation - or do you know this for sure (eg. from sway's source), that in sway for sway ipc -> SUBSCRIBE the sway ipc -> EVENTS -> workspace events (that we are reacting for now) are getting sent so prohibitively late (eg. after already rendering or such) during sway's internal processing of a workspace change

The IPC message for workspace changes will be sent during sway's next mainloop event processing, while the actual workspace change happens during the current loop - although I believe the rendering is still delayed until the next scheduled frame deadline (I didn't verify that, but let's assume it's true because if it's false then there is no way to get in the same frame as the switch).  Note variable refresh rate monitors might make this worse by more commonly being immediately ready for a frame - I don't have one and haven't looked at that code.

In that case, if the input event happens early enough in a frame, then your process will have enough time to see the event and send an attach/commit that sway will read before rendering the next frame.  However, this relies on there being enough time between the incoming event and the next frame that sway gets your commit.  A 60Hz display has a 16ms frame time, and the delay between command start and IPC send took 4ms on the system I checked (measured via `strace -ttt`, which might be pessimistic because it adds overhead).  Assuming that's accurate, you'd be guaranteed to miss the deadline about 1/4 of the time even if your process responded instantly, because sway would start doing rendering in the same iteration that sends the IPC message out.

Even if the IPC gets sent in a loop prior to the rendering step, it's likely that multiple processes will be woken up by the workspace change: clients on the new workspace will get frame callbacks, and other IPC listeners will get the event too.  If you're not scheduled first in line, you will likely need to add another 1-2ms for either scheduling delays or the latency to wake up another core from a sleep state before your process can even start running.  You can avoid this delay by asking for real-time scheduling like audio applications do, but that doesn't feel like the best solution to me.

In summary: if you respond instantly and if my 4ms latency measurement was overly high, you might be able to get your change in the same frame 80-90% of the time.  That can still be visually annoying.

Sending both events from your wallpaper app makes all of these delays work in your favor instead of against you: your process will likely send both events before sway gets scheduled, and so they will both be handled in sway's next mainloop iteration.

**Comment by zDEFz** [comment]
Putting this from rtings.com

```
Refresh rate 	Time between frames 	Minimum input lag
60Hz 	               16.67 ms 	                        8.33 ms
120Hz 	                8.33 ms 	                        4.17 ms
144Hz 	                6.94 ms 	                        3.47 ms
165Hz 	                6.06 ms 	                        3.03 ms
240Hz 	                4.17 ms 	                        2.09 ms
360Hz 	                2.78 ms 	                        1.39 ms
```

Most gamers are on 240Hz now, including me.

**Comment by gergo-salyi** [comment]
> The IPC message for workspace changes will be sent during sway's next mainloop event processing, while the actual workspace change happens during the current loop [...]

Wow, thank you so much! This is useful to know. I will try this other design for our wallpaper setting then in the near future.

**Comment by zDEFz** [comment]
@gergo-salyi I managed to brew my own solution and I don't see it flicker not even on my 240Hz screen.

Demonstration: 
```bash
#!/bin/bash

while true; do
    # Switch to workspace 1
    swaymsg workspace 1

    # Start mpvpaper for the desired output and enable IPC. The wallpaper is set to the first image.
    mpvpaper -o "input-ipc-server=/tmp/mpv-socket" 'DP-2' /home/blu/.config/sway/wallpaper/DP-2/1.png &

    # Use a brief sleep to allow mpvpaper to initialize. This can be adjusted if needed.
    sleep 0.5

    # Load the second image into the same mpvpaper instance using IPC
    echo 'loadfile /home/blu/.config/sway/wallpaper/DP-2/2.png' | socat - /tmp/mpv-socket

    # Switch to workspace 3
    swaymsg workspace 3
done
```

**Comment by zDEFz** [comment]
Here you have a full script that does everything.
And theres no flicker

```bash

#!/bin/bash

mpvpaper -f -o "input-ipc-server=/tmp/mpv-sock" -n 10 'DP-2' /home/blu/.config/sway/wallpaper/DP-2/a.png &

# Function to load the wallpaper based on the workspace number
load_wallpaper() {
    workspace_num=$1
    if (( $workspace_num == 1 )); then
        echo 'loadfile /home/blu/.config/sway/wallpaper/DP-2/a.png' | socat - /tmp/mpv-sock
    fi
    if (( $workspace_num == 3 )); then
        echo 'loadfile /home/blu/.config/sway/wallpaper/DP-2/b.png' | socat - /tmp/mpv-sock
    fi
}

while true; do 
    # Subscribe to workspace change events
    swaymsg -t subscribe '["workspace"]' | while read -r event; do
        # Get the current workspace number
        workspace_num=$(swaymsg -t get_workspaces | jq '.[] | select(.focused == true).num')

        # Load the appropriate wallpaper
        load_wallpaper $workspace_num
    done

done

```

**Comment by zDEFz** [comment]
I think I make a new repo for an appropriate script that maintains more.

**Comment by danieldg** [comment]
@zDEFz you are measuring a different timing than the one being discussed in this issue.  Your ` # Start timing for wallpaper change` timestamp is after the delays that I was discussing, so the fact that the timing starting there is low is nice, but not a solution.  Your video also doesn't include a window on either workspace, which means you can't actually see when sway changes the workspace - only when swaybar updates.  The swaybar update is also triggered by IPC, and even then, if you go through your video frame by frame you can find one where swaybar has updated and the background is still the color associated with the prior desktop.

**Comment by zDEFz** [comment]
> if you go through your video frame by frame you can find one where swaybar has updated and the background is still the color associated with the prior desktop.

I know. And this isn't swaybar. This is i3blocks.
Look, the default performance of this package here is so abysmal that you get flicker. With the difference in what I provided is that you don't.
We don't talk frame perfect, but imperceptible.

Stuff did not work before, stuff works now.
If you want better, then you may wanna add feature-support to swaywm.

**Comment by gergo-salyi** [comment]
> Demonstration in super slo-mo (from iPhone) - realtime maybe 4 seconds?

@zDEFz Can you please re-test this as the following:
 - Have workspace 1 and 2 with very different wallpapers
 - Have a full-screen non-transparent window on workspace 1.
 - Have a full-screen highly transparent window on workspace 2.
 - Now do the test switching from workspace 1 to workspace 2 - a good result ("no flicker") should mean that the highly transparent window on workspace 2 does not get an initial 1 frame with workspace 1's wallpaper behind it.

Unless you have objections this should be the testing methodology for this issue. Can you please post here your observations testing this way?

**Comment by danieldg** [comment]
If you can't get OBS to work, you can try wf-recorder (from wayfire).  I wasn't trying to scold you (sorry if that's how it came across), I was trying to make sure you were fixing the same problem that the original poster was discussing instead of a related one (I read your "stuff did not work before" statement as implying that you had a different problem that you solved).

**Comment by zDEFz** [comment]
> > Demonstration in super slo-mo (from iPhone) - realtime maybe 4 seconds?
> 
> @zDEFz Can you please re-test this as the following:
> 
>     * Have workspace 1 and 2 with very different wallpapers
> 
>     * Have a ~full-screen~ [EDIT] single non-transparent window on workspace 1.
> 
>     * Have a ~full-screen~ [EDIT] single highly transparent window on workspace 2.
> 
>     * Now do the test switching from workspace 1 to workspace 2 - a good result ("no flicker") should mean that the highly transparent window on workspace 2 does not get an initial 1 frame with workspace 1's wallpaper behind it.
> 
> 
> Unless you have objections this should be the testing methodology for this issue. Can you please post here your observations testing this way?

Your instruction is not related to the wlr surface = background
Reading your instruction, this is general sway question.
Therefore, please try making more specific instruction.

> If you can't get OBS to work, you can try wf-recorder (from wayfire). I wasn't trying to scold you (sorry if that's how it > came across), I was trying to make sure you were fixing the same problem that the original poster was discussing instead > of a related one (I read your "stuff did not work before" statement as implying that you had a different problem that you > solved).

What is this, I can only record one display at a time?
How am I supposed to show multi-display functionality with that?

**Comment by danieldg** [comment]
> Therefore, please try making more specific instruction, related to our topic.

Those instructions are how to produce the issue described in the first post in this issue.  If you have a different flicker issue, then it's likely unrelated (and perhaps solved by your code).

> Did you perhaps mean that I am to set opacity level for the surface layer shell for two of the workspaces?

No.  Opacity should be set on the window above the background.  As I understand it, this is a situation that demonstrates the problem:
- Workspace 1 has a light color background
- An opaque window on workspace 1 is opened, covering the light background.  This window might be dark.
- Workspace 2 has been configured with a dark background
- Either no windows are open on workspace 2, or a window with a transparent background is present.
- Switching from workspace 1 to workspace 2 should *not* cause a frame with a light background to be visible.

In practice, there is almost always going to be a frame where the light background is visible *unless* the background change gets sent to sway along with the workspace change (instead of in reaction to the change).  If you can show this is not true in practice, then that's great (less work for everyone).

**Comment by gergo-salyi** [comment]
> Your instruction is not related to the wlr surface = background
> Reading your instruction, this is general sway question.
> I believe in you making a mistake here.
> Therefore, please try making more specific instruction, related to our topic.

@zDEFz I mean that test this the following way please:
- Open up an image editor (eg. gimp), make a ***white*** wallpaper to your screens resolution, save it as `/home/blu/.config/sway/wallpaper/DP-2/1.png` or whatever you have.
- Now you have a white wallpaper for workspace 1.
- Same way make make a ***black*** wallpaper to your screens resolution, save it as `/home/blu/.config/sway/wallpaper/DP-2/2.png` or whatever you have.
- Now you have a black wallpaper for workspace 2.
- So now you have 2 very different wallpapers for workspace 1 and 2.
- Put eg. a browser as a single window to workspace 1
- Now you have a single non-transparent window on workspace 1
- Put eg. a terminal-emulator with window transparency enabled (eg alacritty with `~/.config/alacritty/alacritty.toml` having `[window] opacity = 0.5`) as a single window to workspace 2
- Now you have a single highly transparent window on workspace 2
- Start your wallpaper program you want to test - let's say do the next steps first with my `multibg-sway` running and then thereafter do the next steps with your shell script running.
- Switch to workspace 1
- Now pay attention, record with obs or your phone, etc.
- Switch to workspace 2

If the issue does not appear the you will see the semi-transparent terminal emulator window of workspace 2 appear on your screen having a black wallpaper (workspace 2's wallpaper) behind it. That's it. This is what I call "no flicker".

If this issue here appears then you will see the semi-transparent terminal emulator window of workspace 2 appear on your screen **BUT** for the first frame it will have a white wallpaper (workspace 1's wallpaper) behind it. Next frame things will settle as they should be and it will finally have a black wallpaper (workspace 2's wallpaper) behind it. But you observed a huge "flicker".

Is it clear this way? 

If you have a different issue please tell me how to reproduce that one because I don't get so far what the issue that you're describing is.

**Comment by zDEFz** [comment]
**> If you have a different issue please tell me how to reproduce that one because I don't get so far what the issue that you're describing is.**

This appears to be the case.
I cannot reproduce the issue anymore.

However, I remember that previously, my wallpaper only started to spawn after the workspace had been switched.
As effect, I could notice a delay with of at least 250ms.
I could see the following - when it happened:

1) Wallpaper on (workspace 1) was shown
2) Switch to workspace 2
3) The wallpaper of the previous workspace is flushed and the current workspace 1 appears black
4) Workspace 2 is shown in i3blocks
5) Workspace 2 does not show a Wallpaper yet
6) Wallpaper is being drawn but it is not ready yet. There are gaps noticeably around the edges, and we can see a rectangle already forming
7) Now the new Wallpaper is finished drawing

Switching back to workspace one, and the same behavior can be made out.

**Comment by gergo-salyi** [comment]
@zDEFz please continue this separate issue of yours at https://github.com/gergo-salyi/multibg-sway/issues/10

**Comment by gergo-salyi** [comment]
> You can get lower latency if you initiate the workspace change from your application instead of reacting to it. This would require something like:
>  - Create an IPC mechanism to your client and a command to send it (this could be `echo cmd | socat stdio UNIX-SENDTO:/path/to/socket` or just executing your normal binary with specific arguments)
>  - Changing all bindings that switch workspaces to instead `exec` the IPC client requesting the change
>  - [...]

@danieldg, or anyone else: I started to code this, but I'm facing a problem: workspace changes initiated by key bindings in sway are easy to intercept - but I have no idea how will I **intercept workspace changes initiated by clicks on the status bar**. This will be a roadblock for the quoted design above. Do you have any clever idea how to solve this?

**Comment by gergo-salyi** [comment]
One really bad idea is to proxy SWAYSOCK, but that's just disgusting.

**Comment by danieldg** [comment]
A slightly less bad idea is to modify the status bar so that it can run an arbitrary command when you click on a workspace instead of just 'switch to this workspace'.  Not sure how hard that is to do on swaybar; it might need a patch if that's the default and there's no way to change it in the config.

What might be harder is handling commands that directly focus windows and only switch workspaces as a side effect: `[con_mark="1"] focus`, `for_window [expr] move container to workspace 4`, or things like `focus_on_window_activation` that activate windows without a command at all.

Overall you might end up with a solution that works perfectly most of the time, and works with a frame or two of lag the rest of the time.  Doing better is probably impossible without patching sway (to support workspace-specific layer shell surfaces or similar).

**Comment by gergo-salyi** [comment]
#### I'm so sorry guys, in short this will stay a wontfix for now. I really tried.

#### Summary: 

I made a preview for sending the `workspace 1`, etc sway commands and the wayland surface commits changing at the wallpaper same time. 

Assumption was that this will always be a race but there is a 0.5 - 1 frame-time window (16 ms at 60 fps or 4 ms at 240 fps) where if these two things arrive to sway in this window then will we no flicker.

I think I have proof that this assumption is sadly wrong. The actual window is much smaller, it's **less then 100 us** (microseconds). If I managed to set the timings right on the spot, I saw 80% no flicker - 10% flicker with wallpaper late 1 frame - and 10% flicker with wallpaper early 1 frame. 

I don't know why, I think it's just how sway schedules its rendering and we can do nothing about it. 

#### Details: 

Things here are measured on my laptop, Intel i5-1035G4 with it's Intel iGPU, at 60 Hz display.

I added a [bpftrace](https://github.com/bpftrace/bpftrace) script (in `scripts/ipc.bt`) to measure the timings at microsecond resolution between of syscall exits on wayland and sway-ipc socket reads and syscall enters on socket writes.

Currently we send the wallpaper change wayland surface commit 150 - 250 us after we receive the workspace change sway ipc event.

As I said I made a preview for sending the `workspace 1`, etc sway commands and the wayland surface commits changing at the wallpaper same time. It's on the `setworkspace` branch now (https://github.com/gergo-salyi/multibg-sway/tree/setworkspace). It works but it's **incomplete** and I cannot complete it because I don't know how to solve problems with:
- intercepting workspace changes initiated by clicks on the status bar
- and a new one: how to know which output the workspace is on (in case of multiple monitors) without asking sway

Nevertheless it can be tested with some difficulty:
Edit sway config `~/.config/sway/config`:
```
# For this test now using Mod4 is a MUST !!!
set $mod Mod4

# bindsym $mod+1 workspace number 1
# bindsym $mod+2 workspace number 2
# bindsym $mod+3 workspace number 3
# bindsym $mod+4 workspace number 4
# bindsym $mod+5 workspace number 5
# bindsym $mod+6 workspace number 6
# bindsym $mod+7 workspace number 7
# bindsym $mod+8 workspace number 8
# bindsym $mod+9 workspace number 9
# bindsym $mod+0 workspace number 10
bindsym $mod+1 nop
bindsym $mod+2 nop
bindsym $mod+3 nop
bindsym $mod+4 nop
bindsym $mod+5 nop
bindsym $mod+6 nop
bindsym $mod+7 nop
bindsym $mod+8 nop
bindsym $mod+9 nop
bindsym $mod+0 nop

```
Compile and run. Set environment variables `MULTIBGSWAY_OUTPUT` to the name of the only output where this works now. Set `MULTIBGSWAY_DELAY` to 0 or the number of **micro**seconds that the artificial added thread::sleep delay of the workspace sway ipc command you want:
```
$ git clone --depth=1 --branch=setworkspace https://github.com/gergo-salyi/multibg-sway
$ cd multibg-sway
$ cargo build --release
$ RUST_LOG=trace MULTIBGSWAY_OUTPUT=HDMI-A-1 MULTIBGSWAY_DELAY=20 target/release/multibg-sway --setworkspaces ~/my_wallpaper_dir
```
Only using `Mod4+1`, etc keys will work, clicking on the status bar wont.

Again as I said If I managed to set the timings right on the spot (in my case `MULTIBGSWAY_DELAY=20` microsecons), **I saw 80% no flicker - 10% flicker with wallpaper late 1 frame - and 10% flicker with wallpaper early 1 frame.** 

I think this is just how sway works and there is nothing to do. I cannot add this feature to a release because of its current limitations - and anyways this sub-100 microsecond precision is impossible to hit reliable, with CPU load and scheduling things just will move around.

I'm sorry, I will stop here now. But if there are any more ideas from anyone, feel free to tell me

**Comment by zDEFz** [comment]
Do you think if you use `max_render_time` in sway to `1` it will get better?

**Comment by gergo-salyi** [comment]
```sh
swaymsg output HDMI-A-1 max_render_time 1
swaymsg max_render_time 1
# master branch
multibg-sway ~/wallpapers
# setworkspace branch
MULTIBGSWAY_OUTPUT=HDMI-A-1 MULTIBGSWAY_DELAY=20 target/release/multibg-sway --setworkspaces ~/wallpapers
```
I tried it with all both the variants from above, didn't help for me at all.

**Comment by zDEFz** [comment]
> ```shell
> swaymsg output HDMI-A-1 max_render_time 1
> swaymsg max_render_time 1
> # master branch
> multibg-sway ~/wallpapers
> # setworkspace branch
> MULTIBGSWAY_OUTPUT=HDMI-A-1 MULTIBGSWAY_DELAY=20 target/release/multibg-sway --setworkspaces ~/wallpapers
> ```
> 
> I tried it with all both the variants from above, didn't help for me at all.

Would been too easy...

**Comment by gergo-salyi** [comment]
NOTE: Project is renamed to multibg-wayland as now we support multiple Wayland compositors (see README).

This fundamental issue still stands, but I may shaved off a few milliseconds latency: for multibg-wayland release 0.2.0 if the  new `--gpu` option is set (not by default) now the wallpapers are actually served from GPU memory (again, see README). On wallpaper switches this avoids going through the limits of CPU memory and PCIe bandwidth. 

Latency gains will be hardware specific and can't test them (theoretical 1 ms minimum for a 4K wallpaper over PCIe 3.0). Don't get your hopes up but maybe helps a little bit.
